cpp/dolfinx/la/MatrixCSR.h

```diff
@@ -53,14 +57,25 @@ void set_csr(U&& data, const V& cols, const V& row_ptr, const W& x,
     // Columns indices for row
     auto cit0 = std::next(cols.begin(), row_ptr[row]);
     auto cit1 = std::next(cols.begin(), row_ptr[row + 1]);
-    for (std::size_t c = 0; c < xcols.size(); ++c)
+    for (std::size_t c = 0; c < nc; ++c)
     {
       // Find position of column index
       auto it = std::lower_bound(cit0, cit1, xcols[c]);
       assert(it != cit1);
       std::size_t d = std::distance(cols.begin(), it);
       assert(d < data.size());
-      data[d] = xr[c];
+      if constexpr (BS > 0)
+      {
+        for (int i = 0; i < BS; ++i)
+          for (int j = 0; j < BS; ++j)
+            data[d * BS * BS + i * BS + j] = xr[(i * nc + c) * BS + j];
```

Comment:
`d * BS * BS + i` can be hoisted outside of the `j` loop. Like wise for `i * nc + c) * BS`.